apps/calendar/js/controllers/notifications.js

```diff
@@ -17,20 +18,33 @@ exports.unobserve = function() {
   messageHandler.responder.off('alarm', exports.onAlarm);
 };
 
-exports.onAlarm = function(alarm) {
+exports.onAlarm = co.wrap(function *(alarm) {
   debug('Will request cpu wake lock...');
   var lock = navigator.requestWakeLock('cpu');
   debug('Received cpu lock. Will issue notification...');
-  return issueNotification(alarm).catch(err => {
-    console.error('controllers/notifications', err.toString());
-  }).then(() => {
+  try {
+    yield issueNotification(alarm);
+  } catch (err) {
+    console.error('[controllers/notifications]', err.toString());
+  } finally {
     // release cpu lock with or without errors
     debug('Will release cpu wake lock...');
     lock.unlock();
-  });
-};
+  }
+
+  var storeFactory = core.storeFactory;
+  var alarmStore = storeFactory.get('Alarm');
+  alarm.fired = true;
+  // Save that we've fired this alarm.
+  yield alarmStore.persist(alarm);
+});
+
```

Comment:
Good thought @jrburke. So the data model for calendar is that we have events which can have one or more occurrences (we call them busytimes). A user can set as many reminders for an occurrence as they want. If you set 2 reminders (say one at 5 minutes and one an hour before) for an occurrence, then we wouldn't want to prevent the 5 minute reminder from firing given that the 1 hour reminder had fired.

As far as cleanup goes, we have this "dependent stores" concept in our calendar data handling. Deleting a calendar event cleanly deletes all of the expanded event occurrences. Deleting an event occurrence, similarly, cleanly also deletes all of the occurrence's alarms.